### PR TITLE
switched from http.client to urllib for fetching external resources t…

### PR DESCRIPTION
**Pull request by smaugs**
…o respect http_proxy environment variables

### CODE REVIEW

File: lib/tools.py [hunk 1]
```diff
@@ -56,32 +57,16 @@ def dt2ts(self, dt):
...
-            conn.request("GET", purl, headers=headers)
-        except Exception as e:
-            logger.warning("Problem fetching {0}: {1}".format(url, e))
-            conn.close()
-            return False
-        resp = conn.getresponse()
+        req = urllib.request.Request(url)
+        req.add_header = ('Accept','text/plain')
+        resp = urllib.request.urlopen(req, timeout=5)
+
```
Comment (ohinckel) [inline lib/tools.py]: This could use the timeout value passed to the function instead of hardcoded 5 secs.

File: lib/tools.py [hunk 2]
```diff
@@ -56,32 +57,16 @@ def dt2ts(self, dt):
...
-        if username and password:
-            headers['Authorization'] = ('Basic '.encode() + base64.b64encode((username + ':' + password).encode()))
-        try:
-            conn.request("GET", purl, headers=headers)
-        except Exception as e:
-            logger.warning("Problem fetching {0}: {1}".format(url, e))
-            conn.close()
-            return False
-        resp = conn.getresponse()
+        req = urllib.request.Request(url)
```
Comment (ohinckel) [inline lib/tools.py]: What about the username and password parameters? They are not included in the request in any way.

File: plugins/prowl/__init__.py [hunk 3]
```diff
@@ -55,10 +52,15 @@ def __call__(self, event='', description='', priority=None, url=None, apikey=Non
...
-            resp = conn.getresponse()
-            conn.close()
+            apiurl='https://api.prowlapp.com/publicapi/add'
+            data = urllib.parse.urlencode(data)
+            data = data.encode('utf-8')
+            req = urllib.request.Request(apiurl, data)
+            req.add_header = ('User-Agent','SmartHome.py')
+            req.add_header = ('Content-Type','application/x-www-form-urlencoded;charset=utf-8')
+
+            logger.debug("lala");
```
Comment (ohinckel) [inline plugins/prowl/__init__.py]: This debugging stuff could be changed to something more interesting or could be removed completely ;-)

**Comment by mknx** [comment]
Hi,

I've used url lib in the beginning. But it have had a memory leak. I'm not sure about the request module. I think it build up on the urllib base module.

Have you monitored the memory consumption over several days. It has to be fix after a few days.

so long

Marcus

**Comment by smaugs** [comment]
Hi,

I've been running smarthome.py now for several days without noticing any problematic increase of memory consumption.

I've also fixed and rewritten the enttec-part of the DMX module - I initially submitted the patch for the enttec part - but for some reason it was not working anymore. It cleared out that there are some problems in connection with the python3 serial-port handling. I've rewritten the part - now it uses byte-arrays instead of strings for using the serial.write function.

For me its working now perfectly.

**Pull request closed without merging**